### PR TITLE
Move system components into their own namespace.

### PR DESCRIPTION
**Pull request by brendandburns**
Mostly works, but elastic search is failing e2e.  Will dig in more and fix, but sending now for parallel feedback on the rest of it.

(meta point, this addons stuff is _gross_ we should really have a dedicated binary to do this)

### CODE REVIEW

File: cluster/addons/cluster-monitoring/google/heapster-controller.yaml [hunk 1]
```diff
@@ -2,7 +2,7 @@ apiVersion: v1
 kind: ReplicationController
 metadata: 
   name: monitoring-heapster-v4
-  namespace: default
+  namespace: system
```
Comment (thockin) [inline cluster/addons/cluster-monitoring/google/heapster-controller.yaml]: I thought we agreed on "kubernetes" as the name?  I don't care much, but I don't want wires crossed.

Comment (dchen1107) [inline cluster/addons/cluster-monitoring/google/heapster-controller.yaml]: I think kubernetes is a better name here.

**Comment by dchen1107** [comment]
How about those master components too, like kube-apiserver, etc. ? They are currently in default namespace, should we move them to the same namespace as those addons?

**Comment by brendandburns** [comment]
@k8s-bot ok to test

File: cluster/addons/cluster-monitoring/google/heapster-controller.yaml (hunk 1 above)
Comment (brendandburns) [inline cluster/addons/cluster-monitoring/google/heapster-controller.yaml]: I don't care much, but I prefer descriptive names.  I agree there could be namespace clashes. 'k8s-system'?

Comment (thockin) [inline cluster/addons/cluster-monitoring/google/heapster-controller.yaml]: "k8s" is too obscure for users.  "kube" is better.  "kube-system" or
"kube-cluster" or "kube-infra".  How about "kube-system"?

**Comment by mbforbes** [comment]
@thockin and @dchen1107 are looking; arbitrarily assigning to @dchen1107

**Comment by mbforbes** [comment]
FYI shippable was unhappy and Jenkins hasn't even bothered to give it the time of day

**Comment by satnam6502** [comment]
This is why I think the e2e tests fail for Elasticsearch.
In the Fluentd config file we use DNS to specify the Elasticsearch service: https://github.com/GoogleCloudPlatform/kubernetes/blob/master/cluster/addons/fluentd-elasticsearch/fluentd-es-image/td-agent.conf

There is a line:
```
host elasticsearch-logging.default
```
which should be changed to:
```
host elasticsearch-logging.system
```
and then a new image should be created.

**Comment by satnam6502** [comment]
Note that in https://github.com/GoogleCloudPlatform/kubernetes/blob/master/cluster/addons/fluentd-elasticsearch/kibana-controller.yaml you should update the DNS name for the Elasticsearch service by adding the appropriate suffix e.g. `http://elasticsearch-logging.system:9200`.

**Comment by satnam6502** [comment]
Ah, Kibana is in the same namespace so things should be fine as they are.

**Comment by mikedanese** [comment]
the default serviceaccount is name spaced to the default namespace which is required by kube-dns (and soon heapster) will the default service account still be attached to containers in other namespaces? @liggitt @thockin

**Comment by liggitt** [comment]
Each namespace gets its own default service account. Containers get the token for their namespace's service account (cross namespace references aren't possible)

**Comment by mikedanese** [comment]
@liggitt thanks for clarification.

**Comment by liggitt** [comment]
If you create a namespace, you should immediately see a service account with an API token in it

**Comment by ixdy** [comment]
Note the Jenkins failures here are real:
```console
../src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/dns.go:193: podClient declared and not used
../src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/dns.go:233: podClient declared and not used
```

**Comment by brendandburns** [comment]
@satnam6502 @bgrant0607 

This mostly works for clean clusters.  There are two things I think we need to do to complete this work:

1) Validate that the update path works correctly
2) All of the secrets (including the kubelet and kube-proxy secrets) are getting generated in the kube-system namespace.  I'm not sure if those secrets are really used, since the cluster seems healthy, but we may need to re-configure the proxy and kubelet to use secrets from that namespace.

If no one can get to it, I will come back to it when I return.

Thanks!

**Comment by satnam6502** [comment]
I can take this up on Monday if that's OK (I am OOO Friday).

**Comment by bgrant0607** [comment]
Looks pretty reasonable so far.

File: cluster/saltbase/salt/kube-addons/kube-addons.sh [hunk 2]
```diff
@@ -152,7 +199,7 @@ while read line; do
   else
     # Set the server to https://kubernetes. Pods/components that
     # do not have DNS available will have to override the server.
-    create-kubeconfig-secret "${token}" "${username}" "https://kubernetes"
+    create-kubeconfig-secret "${token}" "${username}" "https://kubernetes.default.svc.cluster.local"
```
Comment (liggitt) [inline cluster/saltbase/salt/kube-addons/kube-addons.sh]: should just need `kubernetes.default`

**Comment by satnam6502** [comment]
I am trying to take over the PR to finish it off and test it. I shall see if I can work out how to "import" the commit somehow into my own PR so we can preserve the history, comments etc.

**Comment by satnam6502** [comment]
This PR is a step towards addressing #9356

**Comment by satnam6502** [comment]
Work continues in PR  #10487

File: cluster/saltbase/salt/kube-addons/kube-addons.sh (hunk 2 above)
Comment (satnam6502) [inline cluster/saltbase/salt/kube-addons/kube-addons.sh]: Done.

**Comment by wojtek-t** [comment]
I'm closing this PR, since it was picked by @satnam6502 in #10487

**Pull request closed without merging**